src/ui/components/Carousel/index.js

```diff
@@ -1,98 +1,60 @@
 /* @flow */
 import classNames from 'classnames';
 import React from 'react';
+import { connect } from 'react-redux';
 import { compose } from 'redux';
-import NukaCarousel from 'nuka-carousel';
 
-import log from 'core/logger';
-import translate from 'core/i18n/translate';
-import { getDirection } from 'core/i18n/utils';
+import { setCarouselOrder } from 'core/reducers/carousels';
 import Card from 'ui/components/Card';
 import CarouselSection from 'ui/components/CarouselSection';
 
 import './styles.scss';
 
 
 type PropTypes = {|
-  i18n: Object,
+  carousels: Object,
+  dispatch: Function,
+  name: string,
+  random?: boolean,
   sections: Array<React.Element<typeof CarouselSection>>,
 |};
 
-type StateTypes = {|
-  hasRendered: boolean,
-|};
-
 export class CarouselBase extends React.Component {
-  constructor(props: PropTypes) {
-    super(props);
-
-    this.state = { hasRendered: false };
-  }
-
-  state: StateTypes;
-
-  // HACK: NukaCarousel uses inline styles to create the carousel effect, which
-  // work fine on the client but cause CSP issues during initial server render.
-  // To get around this, we re-render the component on the client side.
-  // We use componentDidMount to cause it to update on the client because this
-  // lifecycle method is only called on the client.
-  // See: https://github.com/mozilla/addons-frontend/issues/3349
   componentDidMount() {
-    if (!this.state.hasRendered) {
-      log.debug('Re-rendering Carousel to avoid CSP issues.');
-      // eslint-disable-next-line react/no-did-mount-set-state
-      this.setState({ hasRendered: true });
+    const {
+      carousels,
+      dispatch,
+      name,
+      random,
+      sections,
+    } = this.props;
+
+    if (!carousels[name]) {
+      dispatch(setCarouselOrder({ name, random, sections }));
```

Comment:
One question - how does the randomization work in terms server-side renders vs client ones?

Reply:
Since the randomization happens in the reducer, the client side should receive the same ordered list of sections (server state being serialized to be passed to the client).

Reply:
Yup, that's the only reason it's stored in the reducer, otherwise this would be way over-engineered 😉 .